# [Online Learning Perceptron From MLWave](https://mlwave.com/online-learning-perceptron/)

* 온라인 학습 퍼셉트론

* 퍼셉트론 : 가장 단순한 인공신경망을 본다. 이 기사는 1943년에 고안된 개념에서 2015년 Kaggle 경진대회로 옮겨간다. NLP 감정 분석 작업에서 하나의 인공 신경 세포가 0.95 AUC를 얻을 수 있음을 보여준다.(영화 리뷰가 긍정인지 부정인지 예측)

McCulloch-Pitts Neuron
인공 신경망의 탄생은 1943 년의 논문 “a Logical Calculus of the Ideas Immanent in Nervous Activity”. 으로 시작되었다. 신경 학자 인 McCulloch는 논리학자인 Pitts라는 두 연구자가 최초의 인공 뉴런을 스케치하기 위해 힘을 합쳤다.


McCulloch는 신경 활동에 ‘all-or-nothing’활동이 있다고 추론했다. 활성화 임계 값에 도달하거나(출력 1) 출력하지 않으면 (출력 0) 신경세포가 발화한다.
피츠 (Pitts)는 그러한 신경 학적 원칙을 사용하여 명제 논리를 포착 할 수있는 잠재력을보고 이해했다.

<img src='https://mlwave.com/wp-content/uploads/2015/01/mccullogpitts.png'>
McCulloch-Pitts 논문에서 뉴런의 회로도, looping 뉴런과 net

새가 씨앗과 같이 작은 물체를 먹으려고 한다. 이 행동을 표로 나타내면 다음과 같다.



|OBJECT|BROWN?|	ROUND?|	EAT?|
|:---|:---:|---:|:---|
|Seed|	1|	1|	1|
|Leaf|	1|	0|	0|
|Golf Ball|	0|	1|	0|
|Key|	0|	0|	0|

McCulloch-Pitts Neuron으로 위 표를 모델링 할 수 있다. 활성화 임계 값을 1.5로 설정하면 "BROWN"및 "ROUND"속성이 모두 충족 될 때만 뉴런이 실행된다. 그러면 입력 합계는 2 (1 + 1)로 1.5의 활성화 임계 값보다 크다.

<img src='https://mlwave.com/wp-content/uploads/2015/01/pitts-neuron.png'>

### Learning

퍼셉트론은 지도학습 분류기의 일종이다. 이전 값에 대한 학습으로 예측을 한다.
내적(dotproduct) 값이 임계 값보다 높거나 낮은지에 따라 초과하면 "1"을 예측하고 미만이면 "0"을 예측한다.

다음 perceptron은 샘플의 레이블을 봅니다. 예측이 정확하다면, 오류는 "0"이며, 가중치만 남겨 둔다. 예측이 틀린 경우 오류는 "-1"또는 "1"이고 퍼셉트론은 다음과 같이 가중치를 업데이트합니다.

`weights[feature_index] += learning_rate * error * feature_value`


## 퍼셉트론
출처 : [퍼셉트론 - 위키백과, 우리 모두의 백과사전](https://ko.wikipedia.org/wiki/%ED%8D%BC%EC%85%89%ED%8A%B8%EB%A1%A0)


인식 심리학자 Frank Rosenblatt가 1957 년에 발명 한 퍼셉트론 알고리즘은 하드웨어로 구현 된 최초의 인공 신경망이었다.

퍼셉트론(perceptron)은 인공신경망의 한 종류로서, 1957년에 코넬 항공 연구소(Cornell Aeronautical Lab)의 프랑크 로젠블라트 (Frank Rosenblatt)에 의해 고안되었다. 이것은 가장 간단한 형태의 피드포워드(Feedforward) 네트워크 - 선형분류기- 으로도 볼 수 있다.

1960년 코넬 항공 연구소의 연구원은 미 해군 연구청의 자금으로 무작위로 400개의 광전지를 하나의 퍼셉트론에 연결하여 "마크 1 퍼셉트론 (Mark 1 perceptron)"이 탄생해 이것으로 기본 이미지 인식이 가능했다.

퍼셉트론이 동작하는 방식은 다음과 같다. 각 노드의 가중치와 입력치를 곱한 것을 모두 합한 값이 활성함수에 의해 판단되는데, 그 값이 임계치(보통 0)보다 크면 뉴런이 활성화되고 결과값으로 1을 출력한다. 뉴런이 활성화되지 않으면 결과값으로 -1을 출력한다.

마빈 민스키와 시모어 페퍼트는 저서 "퍼셉트론"에서 단층 퍼셉트론은 XOR 연산이 불가능하지만, 다층 퍼셉트론으로는 XOR 연산이 가능함을 보였다.

#### 가중치
퍼셉트론은 들어오는 연결에 가중치를 할당하여 작동한다. McCulloch-Pitts Neuron을 사용하여 들어오는 연결에서 값의 합계를 구하고 다음 특정 임계 값보다 높거나 낮은 지 확인했다. 내적을 구하는 대신 퍼셉트론을 사용했다. 여기에서는 들어오는 각 값에 가중치를 곱해서 합계를 구했다. 
`sum: (value1 * weight1) + (value2 * weight2)`

## 온라인 학습 Perceptron in Python
파이썬으로 위의 퍼셉트론 알고리즘을 구현하며 스크립트가 PyPy(3-4배의 속도 향상)에서 실행되므로 표준 라이브러리를 사용하므로 Kaggle 포럼에서 처음 발견 된 tinrtgu의 온라인 로지스틱 회귀 스크립트에서 큰 영감을 얻었다고 한다. 
* [Display Advertising Challenge - Kaggle, Beat the benchmark with less then 200MB of memory.](https://www.kaggle.com/c/criteo-display-ad-challenge/discussion/10322)


## 온라인 학습
퍼셉트론은 온라인 학습이 가능합니다(한 번에 하나씩 샘플을 통해 학습). 메모리에 전체 데이터 세트가 필요하지 않으므로 더 큰 데이터 세트에 유용합니다. 여기에서는 Vowpal Wabbit에서 온라인 학습 코드를 가져왔다.
* [Vowpal Wabbit (Fast Learning)](http://hunch.net/~vw/)

## 해싱 트릭
벡터화 해싱 트릭은 Vowpal Wabbit(John Langford)에서 시작되었다. 이 트릭은 퍼셉트론으로 들어오는 연결 수를 고정 된 크기로 설정한다. 고정 된 크기보다 낮은 숫자로 모든 원시 피처를 해싱한다.



* 소스코드 출처
    * https://github.com/MLWave/online-learning-perceptron
    * https://github.com/MLWave/online-learning-perceptron/blob/master/perceptron.py
    * 위 소스코드를 Python3에서 실행되도록 일부 수정하였다.

* 위 코드에서는 pandas, numpy, scipy같은 라이브러리를 사용하지 않고 파이썬에 내장 된 기능만을 사용해 학습하도록 한다.

In [1]:
import re
import random
from math import exp, log
from datetime import datetime
from operator import itemgetter

In [2]:
def clean(s):
    """
        Returns a cleaned, lowercased string
        텍스트 데이터를 정제하고 소문자로 변환해 준다.
    """
    return " ".join(re.findall(r'\w+', s, flags = re.UNICODE)).lower()

In [3]:
def get_data_tsv(loc_dataset,opts):
    """
    Running through data in an online manner
    Parses a tsv file for this competition and yields label, identifier and features
    output:
            label: int, The label / target (set to "1" if test set)
            id: string, the sample identifier
            features: list of tuples, in the form [(hashed_feature_index,feature_value)]
            
    온라인 학습 방법을 통해 데이터를 실행한다.
    tsv파일을 통해 레이블, identifier, 피처(특성)를 파싱한다.
    결과물:
        label : int, 레이블 / 대상 (테스트 집합 인 경우 "1"로 설정)
        id : 문자열, 샘플 식별자
        features : [(hashed_feature_index, feature_value)] 형식의 튜플 목록
    """
    for e, line in enumerate(open(loc_dataset,"rb")):
        if e > 0:
            r = line.decode('utf-8').strip().split("\t")
            id = r[0]

            if opts["clean"]:
                try:
                    r[2] = clean(r[2])
                except:
                    r[1] = clean(r[1])

            if len(r) == 3: #train set
                features = [(hash(f)%opts["D"],1) for f in r[2].split()]
                label = int(r[1])
            else: #test set
                features = [(hash(f)%opts["D"],1) for f in r[1].split()]
                label = 1

            if opts["2grams"]:
                for i in range(len(features)-1):
                    features.append(
                        (hash(str(features[i][0])+str(features[i+1][0]))%opts["D"],1))
            yield label, id, features

In [4]:
def dot_product(features,weights):
    """
    Calculate dot product from features and weights
    input:
            features: A list of tuples [(feature_index,feature_value)]
            weights: the hashing trick weights filter, note: length is max(feature_index)
    output:
            dotp: the dot product
    피처(특성)과 가중치로부터 내적을 구한다.
    """
    dotp = 0
    for f in features:
        dotp += weights[f[0]] * f[1]
    return dotp

In [5]:
def train_tron(loc_dataset,opts):
    start = datetime.now()
    print("\nPass\t\tErrors\t\tAverage\t\tNr. Samples\tSince Start")

    # 가중치 초기화
    if opts["random_init"]:
        random.seed(3003)
        weights = [random.random()] * opts["D"]
    else:
        weights = [0.] * opts["D"]

    # Running training passes
    # 학습 실행
    for pass_nr in range(opts["n_passes"]):
        error_counter = 0
        for e, (label, id, features) in enumerate( get_data_tsv(loc_dataset,opts) ):
            
            # 퍼셉트론은 지도학습 분류기의 일종이다. 
            # 이전 값에 대한 학습으로 예측을 한다. 
            # 내적(dotproduct) 값이 임계 값보다 높거나 낮은지에 따라 
            # 초과하면 "1"을 예측하고 미만이면 "0"을 예측한다.
            dp = dot_product(features, weights) > 0.5
            
            # 다음 perceptron은 샘플의 레이블을 봅니다. 
            # 예측이 정확하다면, 오류는 "0"이며, 가중치만 남겨 둔다. 
            # 예측이 틀린 경우 오류는 "-1"또는 "1"이고 퍼셉트론은 다음과 같이 가중치를 업데이트합니다.
            # error is 1 if misclassified as 0, error is -1 if misclassified as 1
            # weights[feature_index] += learning_rate * error * feature_value
            error = label - dp 
        
            if error != 0:
                error_counter += 1
                # Updating the weights
                for index, value in features:
                    weights[index] += opts["learning_rate"] * error * log(1.+value)

        #Reporting stuff
        print("%s\t\t%s\t\t%s\t\t%s\t\t%s" % ( \
            pass_nr+1,
            error_counter,
            round(1 - error_counter /float(e+1),5),
            e+1,datetime.now()-start))

        #Oh heh, we have overfit :)
        if error_counter == 0 or error_counter < opts["errors_satisfied"]:
            print("%s errors found during training, halting"%error_counter)
            break
    return weights

In [6]:
def train_tron(loc_dataset,opts):
    start = datetime.now()
    print("\nPass\t\tErrors\t\tAverage\t\tNr. Samples\tSince Start")

    #Initializing the weights
    if opts["random_init"]:
        random.seed(3003)
        weights = [random.random()] * opts["D"]
    else:
        weights = [0.] * opts["D"]

    #Running training passes
    for pass_nr in range(opts["n_passes"]):
        error_counter = 0
        for e, (label, id, features) in enumerate( get_data_tsv(loc_dataset,opts) ):

            dp = dot_product(features, weights) > 0.5
            error = label - dp # error is 1 if misclassified as 0, error is -1 if misclassified as 1

            if error != 0:
                error_counter += 1
                # Updating the weights
                # 가중치 업데이트
                for index, value in features:
                    weights[index] += opts["learning_rate"] * error * log(1.+value)

        # Reporting stuff
        print("%s\t\t%s\t\t%s\t\t%s\t\t%s" % (
            pass_nr+1,
            error_counter,
            round(1 - error_counter /float(e+1),5),e+1,datetime.now()-start))

        # Oh heh, we have overfit :)
        # 오버피팅 되면 엄춘다.
        if error_counter == 0 or error_counter < opts["errors_satisfied"]:
            print("%s errors found during training, halting"%error_counter)
            break
    return weights


In [7]:
def test_tron(loc_dataset,weights,opts):
    """
        output:
                preds: list, a list with [id,prediction,dotproduct,0-1normalized dotproduct]
    """
    start = datetime.now()
    print("\nTesting online\nErrors\t\tAverage\t\tNr. Samples\tSince Start")
    preds = []
    error_counter = 0
    for e, (label, id, features) in enumerate( get_data_tsv(loc_dataset,opts) ):

        dotp = dot_product(features, weights)
        # 내적이 0.5보다 크다면 긍정으로 예측한다.
        dp = dotp > 0.5
        if dp > 0.5: # we predict positive class
            preds.append( [id, 1, dotp ] )
        else:
            preds.append( [id, 0, dotp ] )

        if label - dp != 0:
            error_counter += 1

    print("%s\t\t%s\t\t%s\t\t%s" % (
        error_counter,
        round(1 - error_counter /float(e+1),5),
        e+1,
        datetime.now()-start))

    # normalizing dotproducts between 0 and 1 
    # TODO: proper probability (bounded sigmoid?), online normalization
    max_dotp = max(preds,key=itemgetter(2))[2]
    min_dotp = min(preds,key=itemgetter(2))[2]
    for p in preds:
        p.append((p[2]-min_dotp)/float(max_dotp-min_dotp)) #appending normalized to predictions

    #Reporting stuff
    print("Done testing in %s"%str(datetime.now()-start))
    return preds

In [8]:
#Setting options
opts = {}
opts["D"] = 2 ** 25
opts["learning_rate"] = 0.1
opts["n_passes"] = 80 # Maximum number of passes to run before halting
opts["errors_satisfied"] = 0 # Halt when training errors < errors_satisfied
opts["random_init"] = False # set random weights, else set all 0
opts["clean"] = True # clean the text a little
opts["2grams"] = True # add 2grams

#training and saving model into weights
%time weights = train_tron("data/labeledTrainData.tsv",opts)


Pass		Errors		Average		Nr. Samples	Since Start
1		5662		0.77352		25000		0:00:16.115041
2		3142		0.87432		25000		0:00:31.234052
3		2304		0.90784		25000		0:00:46.287895
4		1743		0.93028		25000		0:01:01.079004
5		1237		0.95052		25000		0:01:15.817920
6		1037		0.95852		25000		0:01:30.555684
7		788		0.96848		25000		0:01:45.169659
8		667		0.97332		25000		0:02:00.072391
9		484		0.98064		25000		0:02:14.725834
10		458		0.98168		25000		0:02:29.367534
11		379		0.98484		25000		0:02:44.166580
12		381		0.98476		25000		0:02:58.766317
13		270		0.9892		25000		0:03:13.331800
14		209		0.99164		25000		0:03:27.830063
15		214		0.99144		25000		0:03:42.439388
16		182		0.99272		25000		0:03:56.961730
17		141		0.99436		25000		0:04:11.550386
18		155		0.9938		25000		0:04:26.064445
19		118		0.99528		25000		0:04:40.758668
20		115		0.9954		25000		0:04:55.301637
21		134		0.99464		25000		0:05:09.882911
22		70		0.9972		25000		0:05:24.386131
23		62		0.99752		25000		0:05:38.981252
24		37		0.99852		25000		0:05:53.853744
25

In [9]:
#testing and saving predictions into preds
%time preds = test_tron("data/testData.tsv",weights,opts)


Testing online
Errors		Average		Nr. Samples	Since Start
12849		0.48604		25000		0:00:14.563494
Done testing in 0:00:14.581813
CPU times: user 14.5 s, sys: 39.3 ms, total: 14.5 s
Wall time: 14.6 s


In [10]:
#writing kaggle submission
with open("data/submit_perceptron.csv","wb") as outfile:
    outfile.write('"id","sentiment"\n'.encode('utf-8'))
    for p in sorted(preds):
        outfile.write("{},{}\n".format(p[0],p[3]).encode('utf-8'))

In [11]:
# 캐글 스코어 0.95338
129/578

0.2231833910034602



### 상위 벤치 마크
포럼에서 현재 가장 좋은 벤치 마크 는 Abhishek 의 로지스틱 회귀 스크립트다. 리더 보드에서 0.95 AUC 를 얻기 위해 학습과 테스트 세트의 메모리 안에서 tfidf-fit_transform을 사용한다. 여기에서는 이 변환을 건너 뛰고 온라인 해시 벡터 라이저를 사용한다.

Abhishek의 스크립트처럼 기능에서 2g("좋지 않은"등)을 생성한다. 2-gram을 단순히 해쉬하고 샘플 벡터에 추가한다.

또, 문자를 소문자로 바꾸고 영숫자가 아닌 것을 제거함으로써 텍스트를 조금 더 빠르게 정제한다.

### 예측
AUC를 최적화하기 위해 "1"또는 "0"예측만 제출하지 않는다. 이것은 약 0.88의 AUC를 줄 것이다. 여기에서는 0과 1 사이에서 정규화 된 dotproduct를 제출한다.

이 스크립트는 Abhishek의 솔루션 0.952 점수와 비교해 본다. 하지만 이 코드는 단지 몇 MB의 메모리를 필요로하며, tfidf 변환을 수행하지 않고 단지 2분 만에 0의 학습오류에 수렴한다다.

텍스트 정제와 2-gram이 없으면 스크립트는 60초 이내에 실행되고 0.93의 점수를 산출한다.


## 결론
"절대적인 지식은 없다. 모든 것은 확률에 의해서만 이루어진다 " - Gödel (1961)

0-1 임계 값 활성화가 적용된 매우 기본적인 퍼셉트론은 NLP 경진대회에서 잘 동작한다. 여기에서는 1957 알고리즘을 전혀 변경하지 않았다. 해싱 트릭을 없애고 비슷하거나 약간 더 좋은 점수를 받을 수 있다.

이 스크립트는 확률을 출력 할 수 없다. dotproduct에 bounded sigmoid를 사용하여 출력의 온라인 정규화 형태를 얻거나 tinrtgu의 스크립트를 더 보완해 볼 수 있을 것이다.

단일 노드 단일 레이어 퍼셉트론은 비선형 함수를 모델링 할 수 없으므로 더 나은 NLP 결과를 얻으려면 FFN(feedforward nets), recurrent nets, self-organizing maps, MLP(Multi-layer Perceptrons), word2vec 및 extreme learning machine (백프로파게이션이 없는 fast ffnets)을 봐야할 것이다.

